kernel/msg_q.c

```diff
@@ -150,13 +149,50 @@ int z_impl_k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout
 			arch_thread_return_value_set(pending_thread, 0);
 			z_ready_thread(pending_thread);
 		} else {
-			/* put message in queue */
 			__ASSERT_NO_MSG(msgq->write_ptr >= msgq->buffer_start &&
 					msgq->write_ptr < msgq->buffer_end);
-			(void)memcpy(msgq->write_ptr, (char *)data, msgq->msg_size);
-			msgq->write_ptr += msgq->msg_size;
-			if (msgq->write_ptr == msgq->buffer_end) {
-				msgq->write_ptr = msgq->buffer_start;
+			if (put_at_back) {
+				/*
+				 * writing a message to the back of the queue is
+				 * simple and effective: copy the message, then
+				 * increment write_ptr.
+				 */
+				(void)memcpy(msgq->write_ptr, (char *)data, msgq->msg_size);
+				msgq->write_ptr += msgq->msg_size;
+				if (msgq->write_ptr == msgq->buffer_end) {
+					msgq->write_ptr = msgq->buffer_start;
+				}
+			} else {
+				/*
+				 * writing a message to the head of the queue is
+				 * less obvious, but it can be achieved by simply
+				 * following these steps:
+				 *
+				 * 1. decrementing the read pointer. This effectively
+				 * opens space for the incoming message in the head of
+				 * the queue.
+				 *
+				 * 2. temporarily matching the write pointer with the
+				 * read pointer. This way the message will be written
+				 * in the recently opened space.
+				 *
+				 * 3. reverting the write pointer after the write to its
+				 * original value. The read pointer, by its turn, should
+				 * be always at the head, so it doesn't need reverting
+				 * because we just wrote to the head.
+				 *
+				 * ...but this is inefficient because we need to go back
+				 * and forth with the value of write_ptr; note that it
+				 * becomes equal to read_ptr just to be restored afterwards.
+				 * A much more efficient solution is to simply write the
+				 * message directly to read_ptr's address after step 1,
+				 * which avoids needing to perform steps 2 and 3 altogether.
```

Comment:
Let's simplify this comment. Perhaps something like ...

				 * writing a message to the front of the queue is
				 * simple and effective: create space for the message
				 * by decrementing read_ptr, then copy the message
				 * to the newly created space

Reply:
alright, I have reduced the message. I still think the full rationale is valuable, though.